src/cli.rs

```diff
@@ -606,77 +579,66 @@ fn open_channel(
-	router: Arc<NetGraphMsgHandler<Arc<dyn chain::Access + Send + Sync>, Arc<FilesystemLogger>>>,
-	channel_manager: Arc<ChannelManager>, payment_storage: PaymentInfoStorage,
-	logger: Arc<FilesystemLogger>,
+fn send_payment<E: EventHandler>(
+	invoice_payer: &InvoicePayer<E>, invoice: &Invoice, payment_storage: PaymentInfoStorage,
 ) {
-	let network_graph = &router.network_graph;
-	let first_hops = channel_manager.list_usable_channels();
-	let payer_pubkey = channel_manager.get_our_node_id();
-
-	let route = router::get_route(
-		&payer_pubkey,
-		&network_graph,
-		&payee,
-		payee_features,
-		Some(&first_hops.iter().collect::<Vec<_>>()),
-		&route_hints,
-		amt_msat,
-		final_cltv,
-		logger,
-		&Scorer::default(),
-	);
-	if let Err(e) = route {
-		println!("ERROR: failed to find route: {}", e.err);
-		return;
-	}
-	let status = match channel_manager.send_payment(&route.unwrap(), payment_hash, &payment_secret)
-	{
+	let status = match invoice_payer.pay_invoice(invoice) {
 		Ok(_payment_id) => {
-			println!("EVENT: initiated sending {} msats to {}", amt_msat, payee);
+			let payee_pubkey = invoice.recover_payee_pub_key();
+			let amt_msat = invoice.amount_milli_satoshis().unwrap();
+			println!("EVENT: initiated sending {} msats to {}", amt_msat, payee_pubkey);
+			print!("> ");
 			HTLCStatus::Pending
 		}
-		Err(e) => {
+		Err(PaymentError::Invoice(e)) => {
+			println!("ERROR: invalid invoice: {}", e);
+			print!("> ");
+			return;
+		}
+		Err(PaymentError::Routing(e)) => {
+			println!("ERROR: failed to find route: {}", e.err);
+			print!("> ");
+			return;
+		}
+		Err(PaymentError::Sending(e)) => {
 			println!("ERROR: failed to send payment: {:?}", e);
+			print!("> ");
 			HTLCStatus::Failed
 		}
 	};
+	let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
+	let payment_secret = Some(invoice.payment_secret().clone());
+
 	let mut payments = payment_storage.lock().unwrap();
 	payments.insert(
 		payment_hash,
 		PaymentInfo {
 			preimage: None,
 			secret: payment_secret,
 			status,
-			amt_msat: MillisatAmount(Some(amt_msat)),
+			amt_msat: MillisatAmount(invoice.amount_milli_satoshis()),
 		},
 	);
 }
 
 fn keysend(
-	payee: PublicKey, amt_msat: u64,
-	router: Arc<NetGraphMsgHandler<Arc<dyn chain::Access + Send + Sync>, Arc<FilesystemLogger>>>,
+	payee_pubkey: PublicKey, amt_msat: u64, network_graph: Arc<NetworkGraph>,
 	channel_manager: Arc<ChannelManager>, payment_storage: PaymentInfoStorage,
-	logger: Arc<FilesystemLogger>,
+	logger: Arc<FilesystemLogger>, scorer: Arc<Mutex<Scorer>>,
 ) {
-	let network_graph = &router.network_graph;
 	let first_hops = channel_manager.list_usable_channels();
 	let payer_pubkey = channel_manager.get_our_node_id();
 
-	let route = match router::get_keysend_route(
+	let payee = Payee::for_keysend(payee_pubkey);
+	let params = RouteParameters { payee, final_value_msat: amt_msat, final_cltv_expiry_delta: 40 };
+
+	let route = match router::find_route(
```

Comment:
Hmmmmmm, we really need to get `InvoicePayer` to support spontaneous payments eventually.